src/Functions/vectorFunctions.cpp

```diff
@@ -0,0 +1,1067 @@
+static const char DIVIDE_NAME[] = "divide";
+
+static constexpr char L1_LABEL[] = "1";
+static constexpr char L2_LABEL[] = "2";
+static constexpr char Linf_LABEL[] = "inf";
+static constexpr char Lp_LABEL[] = "p";
+
+/// str starts from the lowercase letter; not constexpr due to the compiler version
+/*constexpr*/ std::string makeFirstLetterUppercase(std::string && str)
+{
+    std::string res(str);
+    res[0] += 'A' - 'a';
+    return res;
+}
+
+template <const char * func_name>
+class FunctionTupleOperator : public TupleIFunction
+{
+public:
+    /// constexpr cannot be used due to std::string has not constexpr constructor in this compiler version
+    static inline auto name = "tuple" + makeFirstLetterUppercase(func_name);
+
+    explicit FunctionTupleOperator(ContextPtr context_) : TupleIFunction(context_) {}
+    static FunctionPtr create(ContextPtr context_) { return std::make_shared<FunctionTupleOperator>(context_); }
+
+    String getName() const override { return name; }
+
+    size_t getNumberOfArguments() const override { return 2; }
+
+    DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
+    {
+        const auto * left_tuple = checkAndGetDataType<DataTypeTuple>(arguments[0].type.get());
+        const auto * right_tuple = checkAndGetDataType<DataTypeTuple>(arguments[1].type.get());
+
+        if (!left_tuple)
+            throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Argument 0 of function {} should be tuples, got {}",
+                            getName(), arguments[0].type->getName());
+
+        if (!right_tuple)
+            throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Argument 1 of function {} should be tuples, got {}",
+                            getName(), arguments[1].type->getName());
+
+        const auto & left_types = left_tuple->getElements();
+        const auto & right_types = right_tuple->getElements();
+
+        Columns left_elements;
+        Columns right_elements;
+        if (arguments[0].column)
+            left_elements = getTupleElements(*arguments[0].column);
+        if (arguments[1].column)
+            right_elements = getTupleElements(*arguments[1].column);
+
+        if (left_types.size() != right_types.size())
+            throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT,
+                            "Expected tuples of the same size as arguments of function {}. Got {} and {}",
+                            getName(), arguments[0].type->getName(), arguments[1].type->getName());
+
+        size_t tuple_size = left_types.size();
+        if (tuple_size == 0)
+            return std::make_shared<DataTypeUInt8>();
+
+        auto func = FunctionFactory::instance().get(func_name, context);
+        DataTypes types(tuple_size);
+        for (size_t i = 0; i < tuple_size; ++i)
+        {
+            try
+            {
+                ColumnWithTypeAndName left{left_elements.empty() ? nullptr : left_elements[i], left_types[i], {}};
```

Comment:
Is there real cases where it's empty?

Reply:
Just copypasted from the `tupleHammingDistance`... Actually, do not know (and I agree -- it is strange if there is a real case). I can delete it from all places. If old tests with `tupleHammingDistance` not fail, then it will be possible to conclude that there are no such cases :)

Reply:
Yes, there are!!! `SELECT tupleHammingDistance(materialize((1, 2)), (1, 4));` -- do not know why, but it fails if you leave only the third part of the ternary operator.